File: src/main/java/dev/cosgy/TextToSpeak/Bot.java
////////////////////////////////////////////////////////////////////////////////
//  Copyright 2021 Cosgy Dev                                                   /
//                                                                             /
//     Licensed under the Apache License, Version 2.0 (the "License");         /
//     you may not use this file except in compliance with the License.        /
//     You may obtain a copy of the License at                                 /
//                                                                             /
//        http://www.apache.org/licenses/LICENSE-2.0                           /
//                                                                             /
//     Unless required by applicable law or agreed to in writing, software     /
//     distributed under the License is distributed on an "AS IS" BASIS,       /
//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied./
//     See the License for the specific language governing permissions and     /
//     limitations under the License.                                          /
////////////////////////////////////////////////////////////////////////////////

package dev.cosgy.TextToSpeak;

import com.jagrosh.jdautilities.commons.waiter.EventWaiter;
import dev.cosgy.TextToSpeak.audio.AudioHandler;
import dev.cosgy.TextToSpeak.audio.Dictionary;
import dev.cosgy.TextToSpeak.audio.PlayerManager;
import dev.cosgy.TextToSpeak.audio.VoiceCreation;
import dev.cosgy.TextToSpeak.gui.GUI;
import dev.cosgy.TextToSpeak.settings.SettingsManager;
import dev.cosgy.TextToSpeak.settings.UserSettingsManager;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.entities.Guild;

import java.util.Locale;
import java.util.Objects;
import java.util.ResourceBundle;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;


public class Bot {
    public static Bot INSTANCE;
    private final EventWaiter waiter;
    private final ScheduledExecutorService threadpool;
    private final BotConfig config;
    private final SettingsManager settings;
    private final ResourceBundle lang;
    private final PlayerManager players;
    private final VoiceCreation voiceCreation;
    private final UserSettingsManager userSettingsManager;
    private final Dictionary dictionary;

    private boolean shuttingDown = false;
    private JDA jda;
    private GUI gui;


    public Bot(EventWaiter waiter, BotConfig config, SettingsManager settings, VoiceCreation voiceCreation, UserSettingsManager userSettingsManager, Dictionary dictionary) {
        this.waiter = waiter;
        this.lang = ResourceBundle.getBundle("lang.yomiage", Locale.JAPAN);
        this.config = config;
        this.settings = settings;
        this.threadpool = Executors.newSingleThreadScheduledExecutor();
        this.players = new PlayerManager(this);
        this.players.init();
        this.voiceCreation = voiceCreation;
        voiceCreation.Init(this);
        this.userSettingsManager = userSettingsManager;
        this.dictionary = dictionary;
    }

    public JDA getJDA() {
        return jda;
    }

    public void setJDA(JDA jda) {
        this.jda = jda;
    }

    public void closeAudioConnection(long guildId) {
        Guild guild = jda.getGuildById(guildId);
        if (guild != null)
            threadpool.submit(() -> guild.getAudioManager().closeAudioConnection());
    }

    public void resetGame() {
        Activity game = config.getGame() == null || config.getGame().getName().toLowerCase().matches("(none|なし)") ? null : config.getGame();
        if (!Objects.equals(jda.getPresence().getActivity(), game))
            jda.getPresence().setActivity(game);
    }

    public void shutdown() {
        if (shuttingDown)
            return;
        shuttingDown = true;
        threadpool.shutdownNow();
        if (jda.getStatus() != JDA.Status.SHUTTING_DOWN) {
            jda.getGuilds().forEach(g ->
            {
                g.getAudioManager().closeAudioConnection();
                AudioHandler ah = (AudioHandler) g.getAudioManager().getSendingHandler();
                if (ah != null) {
                    ah.stopAndClear();
                    ah.getPlayer().destroy();
                }
            });
            jda.shutdown();
        }
        if (gui != null)
            gui.dispose();
        System.exit(0);
    }

    public void setGUI(GUI gui) {
        this.gui = gui;
    }

    public ResourceBundle GetLang(){
        return lang;
    }

    public SettingsManager getSettingsManager() {
        return settings;
    }

    public PlayerManager getPlayerManager() {
        return players;
    }

    public EventWaiter getWaiter() {
        return waiter;
    }

    public BotConfig getConfig() {
        return config;
    }

    public ScheduledExecutorService getThreadpool() {
        return threadpool;
    }

    public VoiceCreation getVoiceCreation(){
        return voiceCreation;
    }

    public UserSettingsManager getUserSettingsManager(){
        return userSettingsManager;
    }

    public Dictionary getDictionary(){return dictionary;}
}

File: src/main/java/dev/cosgy/textToSpeak/Bot.kt
//////////////////////////////////////////////////////////////////////////////////////////
//  Copyright 2023 Cosgy Dev                                                             /
//                                                                                       /
//     Licensed under the Apache License, Version 2.0 (the "License");                   /
//     you may not use this file except in compliance with the License.                  /
//     You may obtain a copy of the License at                                           /
//                                                                                       /
//        http://www.apache.org/licenses/LICENSE-2.0                                     /
//                                                                                       /
//     Unless required by applicable law or agreed to in writing, software               /
//     distributed under the License is distributed on an "AS IS" BASIS,                 /
//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.          /
//     See the License for the specific language governing permissions and               /
//     limitations under the License.                                                    /
//////////////////////////////////////////////////////////////////////////////////////////
package dev.cosgy.textToSpeak

import com.jagrosh.jdautilities.commons.waiter.EventWaiter
import dev.cosgy.textToSpeak.audio.*
import dev.cosgy.textToSpeak.audio.Dictionary
import dev.cosgy.textToSpeak.gui.GUI
import dev.cosgy.textToSpeak.settings.SettingsManager
import dev.cosgy.textToSpeak.settings.UserSettingsManager
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import org.apache.commons.io.FileUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.function.Consumer
import kotlin.system.exitProcess

class Bot(val waiter: EventWaiter, val config: BotConfig, val settingsManager: SettingsManager) {
    val threadpool: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private val lang: ResourceBundle = ResourceBundle.getBundle("lang.yomiage", Locale.JAPAN)
    val playerManager: PlayerManager = PlayerManager(this)
    val voiceCreation: VoiceCreation
    val userSettingsManager: UserSettingsManager
    private val aloneInVoiceHandler: AloneInVoiceHandler
    var log: Logger = LoggerFactory.getLogger(this.javaClass)
    var dictionary: Dictionary? = null
        private set
    private var shuttingDown = false
    var jda: JDA? = null
    private var gui: GUI? = null

    init {
        playerManager.init()
        voiceCreation = VoiceCreation(this)
        userSettingsManager = UserSettingsManager()
        aloneInVoiceHandler = AloneInVoiceHandler(this)
        aloneInVoiceHandler.init()
    }

    fun readyJDA() {
        dictionary = Dictionary.getInstance(this)
    }

    fun closeAudioConnection(guildId: Long) {
        val guild = jda!!.getGuildById(guildId)
        if (guild != null) threadpool.submit { guild.audioManager.closeAudioConnection() }
    }

    fun resetGame() {
        val game = if (config.game == null || config.game!!.name.lowercase(Locale.getDefault())
                .matches("(none|なし)".toRegex())
        ) null else config.game
        if (jda!!.presence.activity != game) jda!!.presence.activity = game
    }

    fun shutdown() {
        if (shuttingDown) return
        shuttingDown = true
        if (jda!!.status != JDA.Status.SHUTTING_DOWN) {
            jda!!.guilds.forEach(Consumer { g: Guild ->
                val am = g.audioManager
                if (am.isConnected) {
                    am.closeAudioConnection()
                    val ah = am.sendingHandler as AudioHandler?
                    if (ah != null) {
                        ah.stopAndClear()
                        ah.player.destroy()
                    }
                }
            })

            // Wait for any remaining tasks to complete before shutting down the thread pool
            threadpool.shutdown()
            try {
                if (!threadpool.awaitTermination(10, TimeUnit.SECONDS)) {
                    threadpool.shutdownNow()
                    if (!threadpool.awaitTermination(10, TimeUnit.SECONDS)) {
                        log.warn("Thread pool did not terminate")
                    }
                }
            } catch (e: InterruptedException) {
                threadpool.shutdownNow()
                Thread.currentThread().interrupt()
                log.warn("Thread pool shutdown was interrupted")
            }
        }

        // Shutdown JDA
        jda?.shutdown()
        try {
            if (!jda!!.awaitShutdown(10, TimeUnit.SECONDS)) {
                log.warn("JDA did not shutdown properly")
            }
        } catch (e: InterruptedException) {
            jda!!.shutdownNow()
            Thread.currentThread().interrupt()
            log.warn("JDA shutdown was interrupted")
        }

        dictionary?.close()
        // Delete temporary files
        try {
            FileUtils.cleanDirectory(File("tmp"))
            FileUtils.cleanDirectory(File("wav"))
            log.info("Deleted temporary files")
        } catch (e: IOException) {
            log.warn("Failed to delete temporary files")
        } catch (e: IllegalArgumentException) {
            log.warn("One or more directory paths were invalid.")
        }

        if (gui != null) gui!!.dispose()
        exitProcess(0)
    }


    fun setGUI(gui: GUI?) {
        this.gui = gui
    }

    fun GetLang(): ResourceBundle {
        return lang
    }

    companion object {
        var INSTANCE: Bot? = null
    }
}